modules/dxp/apps/portal-workflow/portal-workflow-kaleo-designer-web/src/main/resources/META-INF/resources/designer/js/definition-builder/diagram-builder/components/sidebar/sections/assignments/select-assignment/Role.js

```diff
@@ -0,0 +1,152 @@
+ * contacting Liferay, Inc. See the License for the specific language governing
+ * permissions and limitations under the License, including but not limited to
+ * distribution rights of the Software.
+ */
+
+import ClayAutocomplete from '@clayui/autocomplete';
+import {useResource} from '@clayui/data-provider';
+import ClayDropDown from '@clayui/drop-down';
+import ClayForm, {ClayInput} from '@clayui/form';
+import React, {useContext, useState} from 'react';
+
+import {headers} from '../../../../../../util/fetchUtil';
+import {DiagramBuilderContext} from '../../../../../DiagramBuilderContext';
+import SidebarPanel from '../../../SidebarPanel';
+
+const Roles = () => {
+	const {setSelectedItem} = useContext(DiagramBuilderContext);
+
+	const [active, setActive] = useState(false);
+	const [filter, setFilter] = useState(true);
+	const [fieldValues, setFieldValues] = useState({id: '', name: ''});
+	const [networkStatus, setNetworkStatus] = useState(4);
+	const {resource} = useResource({
+		fetchOptions: {
+			headers: {
+				...headers,
+				'accept': `application/json`,
+				'x-csrf-token': Liferay.authToken,
+			},
+		},
+		fetchPolicy: 'cache-first',
+		link: `${window.location.origin}/o/headless-admin-user/v1.0/roles`,
+		onNetworkStatusChange: setNetworkStatus,
+	});
+
+	const initialLoading = networkStatus === 1;
+	const loading = networkStatus < 4;
+	const error = networkStatus === 5;
+
+	const handleInputFocus = () => {
+		setFilter(fieldValues.name === '');
+		setActive(true);
+	};
+
+	const handleItemClick = (item) => {
+		setFieldValues({id: item.id, name: item.name});
+		setActive(false);
+		updateSelectedItem(item);
+	};
+
+	const updateSelectedItem = (role) => {
+		setSelectedItem((previousValue) => ({
+			...previousValue,
+			data: {
+				...previousValue.data,
+				assignments: {assignmentType: ['roleId'], roleId: role.id},
+			},
+		}));
+	};
+
+	return (
+		<SidebarPanel panelTitle={Liferay.Language.get('select-role')}>
+			<ClayForm.Group>
+				<ClayAutocomplete>
+					<label htmlFor="role-name">
+						{Liferay.Language.get('role')}
+
+						<span className="ml-1 mr-1 text-warning">*</span>
+					</label>
+
+					<ClayAutocomplete.Input
+						id="role-name"
+						onChange={(event) => {
+							event.persist();
+							setFilter(true);
+							setFieldValues((previousValues) => ({
+								...previousValues,
+								name: event.target.value,
+							}));
+						}}
+						onFocus={() => handleInputFocus()}
+						placeholder="Search"
+						value={fieldValues.name}
+					/>
+
+					<ClayAutocomplete.DropDown
+						active={(!!resource && active) || initialLoading}
+						closeOnClickOutside
+						onSetActive={setActive}
+					>
+						<ClayDropDown.ItemList>
+							{(error || (resource && resource.error)) && (
+								<ClayDropDown.Item className="disabled">
+									{Liferay.Language.get('no-results-found')}
+								</ClayDropDown.Item>
+							)}
+
+							{!error &&
```

Comment:
@cgmonte I think you could move this logic to display the component to a constant, to avoid cluttering the JSX and make it more clean :)

Reply:
Just did. And took the chance to extract the onChange logic from JSX as well. Thanks.